### PR TITLE
add an implementation of ghost gradients

### PR DESCRIPTION
**Pull request by g-w1**
Implements Ghost Gradients from https://transformer-circuits.pub/2024/jan-update/index.html#dict-learning-resampling

Thanks @woog97 for pair programming with me on this!

### CODE REVIEW

File: training.py [hunk 1]
```diff
@@ -125,10 +181,14 @@ def trainSAE(
...
 
+    num_samples_since_activated = t.zeros(dictionary_size, dtype=int).to(device) if ghost_threshold is not None else None # how many samples since each neuron was last activated?
+
+    assert not (ghost_threshold is not None and resample_steps is not None)
+
     # set up optimizer and scheduler
     optimizer = ConstrainedAdam(ae.parameters(), ae.decoder.parameters(), lr=lr)
     def warmup_fn(step):
-        if step % resample_steps < warmup_steps:
+        if resample_steps and step % resample_steps < warmup_steps:
```
Comment (g-w1) [inline training.py:191]: This disables learning rate warmup when using ghost grads instead of resampling.
Would it instead make sense to just do a singular warmup at the start of training if we are using Ghost Grads (rather than disabling it, since [this post](https://www.lesswrong.com/posts/fifPCos6ddsmJYahD/my-best-guess-at-the-important-tricks-for-training-1l-saes#3__Periodic_LR_warmup_) says it's useful)?

Comment (saprmarks) [reply training.py:190]: I think we should do a warmup at the start of training -- even with ghost grads, I expect that the first few gradient steps will kill a bunch of neurons without warming up first.

**Comment by g-w1** [comment]
Sounds good; this matches what I'm seeing in my experiments
